### ISSUE TITLE
save / restore from model checkpoint

### ISSUE THREAD

**Issue by hrbigelow** [google-deepmind/penzai#111]
Hi,

I wondered, if I have a custom designed Penzai model and I'd like to create a checkpoint using orbax.checkpoint, then restore it later, what is the right approach?  Should I attemtp to get orbax to save named axis names?  Or, just save raw Jax.Arrays, and then reconstruct the names later?  Any code snippets would be greatly appreciated!

In the docs, it states:

https://penzai.readthedocs.io/en/stable/guides/howto_reference.html#checkpointing-model-parameters

> Penzai does not include any Penzai-specific checkpointing utilities. However, Penzai is compatible with any PyTree-based JAX checkpointing system, as long as you first extract the parameters using pz.unbind_parameters(model, freeze=True). One checkpointing library that is currently well supported in JAX is [orbax.checkpoint](https://orbax.readthedocs.io/en/latest/orbax_checkpoint_101.html).

So, I was able to successfully checkpoint the parameter values using:

```python
def save_checkpoint(model: pz.nn.Layer, ckpt_path: str):
    _, params = pz.unbind_params(model, freeze=False)
    ckptr = ocp.AsyncCheckpointer(ocp.StandardCheckpointHandler())
    ckptr.save(ckpt_path, ocp.args.StandardSave(params), force=False)
    ckptr.wait_until_finished()
```

However, attempting to reconstruct a model from these saved parameters I cannot figure out yet.

```python
# This doesn't work - 
def load_checkpoint(model: pz.nn.Layer, ckpt_path: str) -> pz.nn.Layer:
    checkpointer = ocp.PyTreeCheckpointer()
    params = checkpointer.restore(ckpt_path)
    model_def, cur_params = pz.unbind_params(model, freeze=True)
    return pz.bind_variables(model_def, params)
```

I thought maybe I'm missing something simple, but can't find it.

Thanks in advance!

**Comment by danieldjohnson** [comment]
Ah, I think the issue might be that orbax can only restore non-builtin pytree types (like NamedArray) if you give it an "item" to tell it what the custom pytree structure is. Does something like this work?

```python
def load_checkpoint(model: pz.nn.Layer, ckpt_path: str) -> pz.nn.Layer:
    checkpointer = ocp.PyTreeCheckpointer()
    model_def, cur_params = pz.unbind_params(model, freeze=True)
    params = checkpointer.restore(ckpt_path, args=ocp.args.PyTreeRestore(item=cur_params))
    return pz.bind_variables(model_def, params)
```

I'm not sure this is the current recommended way to do things in orbax; it's based on some docs [here](https://orbax.readthedocs.io/en/latest/guides/checkpoint/checkpointing_pytrees.html#restoring) and [here](https://orbax.readthedocs.io/en/latest/api_reference/checkpoint.handlers.html#orbax.checkpoint.handlers.PyTreeCheckpointHandler.restore).

**Comment by hrbigelow** [comment]
Thanks @danieldjohnson.  Apologies, I've been too busy to properly respond.  For the moment we found something that works:

[https://github.com/Astera-org/simplexity/blob/main/simplexity/persistence/local_penzai_persister.py](https://github.com/Astera-org/simplexity/blob/main/simplexity/persistence/local_penzai_persister.py
)

Also I asked the folks at orbax - they may look into it:  [here](https://github.com/google/orbax/issues/1833)

From what I understand, the main thing the client needs to do to use orbax with new types is to write code that 1) splits the new type into pure jax.Array plus POD metadata (they recommended a JSON sting), and 2) the reverse of that.
